### PR TITLE
[SPARK-25306][SQL] Avoid skewed filter trees to speed up `createFilter` in ORC

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

In both ORC data sources, `createFilter` function has exponential time complexity due to its skewed filter tree generation. This PR aims to improve it by using new `buildTree` function.

**REPRODUCE**
```scala
// Create and read 1 row table with 1000 columns
sql("set spark.sql.orc.filterPushdown=true")
val selectExpr = (1 to 1000).map(i => s"id c$i")
spark.range(1).selectExpr(selectExpr: _*).write.mode("overwrite").orc("/tmp/orc")
print(s"With 0 filters, ")
spark.time(spark.read.orc("/tmp/orc").count)

// Increase the number of filters
(20 to 30).foreach { width =>
  val whereExpr = (1 to width).map(i => s"c$i is not null").mkString(" and ")
  print(s"With $width filters, ")
  spark.time(spark.read.orc("/tmp/orc").where(whereExpr).count)
}
```

**RESULT**
```scala
With 0 filters, Time taken: 653 ms                                              
With 20 filters, Time taken: 962 ms
With 21 filters, Time taken: 1282 ms
With 22 filters, Time taken: 1982 ms
With 23 filters, Time taken: 3855 ms
With 24 filters, Time taken: 6719 ms
With 25 filters, Time taken: 12669 ms
With 26 filters, Time taken: 25032 ms
With 27 filters, Time taken: 49585 ms
With 28 filters, Time taken: 98980 ms    // over 1 min 38 seconds
With 29 filters, Time taken: 198368 ms   // over 3 mins
With 30 filters, Time taken: 393744 ms   // over 6 mins
```

**AFTER THIS PR**
```scala
With 0 filters, Time taken: 774 ms
With 20 filters, Time taken: 601 ms
With 21 filters, Time taken: 399 ms
With 22 filters, Time taken: 679 ms
With 23 filters, Time taken: 363 ms
With 24 filters, Time taken: 342 ms
With 25 filters, Time taken: 336 ms
With 26 filters, Time taken: 352 ms
With 27 filters, Time taken: 322 ms
With 28 filters, Time taken: 302 ms
With 29 filters, Time taken: 307 ms
With 30 filters, Time taken: 301 ms
```

## How was this patch tested?

Pass the Jenkins with newly added test cases.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this PR, @gatorsmile and @cloud-fan ?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala [hunk 1]
```diff
@@ -55,19 +59,52 @@ import org.apache.spark.sql.types._
...
+    .expireAfterAccess(cacheExpireTimeout, TimeUnit.SECONDS)
+    .build(
+      new CacheLoader[FilterWithTypeMap, Option[Builder]]() {
+        override def load(typeMapAndFilter: FilterWithTypeMap): Option[Builder] = {
+          buildSearchArgument(
+            typeMapAndFilter.typeMap, typeMapAndFilter.filter, SearchArgumentFactory.newBuilder())
+        }
+      })
+
+  private def getOrBuildSearchArgumentWithNewBuilder(
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: @xuanyuanking . This already reuses `cacheExpireTimeout`.

For the cache value, `SearchArgument`, `SearchArgumentFactory` and `Builder` are different. 
- Here, they comes from `org.apache.hadoop.hive.ql.io.sarg.*`.
- There, they comes from `org.apache.orc.storage.ql.io.sarg.*`.

The only exception I made is `FilterWithTypeMap`. I wanted to keep them separately since it's also related to cache key.

**Comment by kiszk** [comment]
General question: Why do we use `time` instead of `entry size` to control cache? I am neutral on this decision. I would like to hear the reason of this decision.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @kiszk .

First, I don't want to hold the memory up after query completion. If we do, it will be a regression. So, I wanted `time` first.

Second, It's difficult to estimate the enough limit for the number of filters.
  - As we know codegen JVM limit issue. There are several attempts to generate a single complex query for wide tables (thousands of columns).
  - Spark's optimizer like `InferFiltersFromConstraints` adds more constraints like 'NotNull(col1)`. Usually, the number of filters becomes double here.
    - Also, it's not a good design if we need to increase this limitation whenever we add a new optimizer like `InferFiltersFromConstraints`.
  - If the limit is too high, we waste the memory. If the limit is small, the eviction will bite us again.

In short, `time` was enough and the simplest for this purpose.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala [hunk 2]
```diff
@@ -55,19 +59,52 @@ import org.apache.spark.sql.types._
...
+      dataTypeMap: Map[String, DataType],
+      expression: Filter): Option[Builder] = {
+    // When `spark.sql.orc.cache.sarg.timeout` is 0, cache is disabled.
+    if (cacheExpireTimeout > 0) {
+      searchArgumentCache.get(FilterWithTypeMap(expression, dataTypeMap))
+    } else {
+      buildSearchArgument(dataTypeMap, expression, SearchArgumentFactory.newBuilder())
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: When we set timeout to zero on the cache, the loaded element can be removed immediately. Maybe we don't need to check timeout like this and we can simplify the code.

**Comment by cloud-fan** [comment]
Do you know why `createFilter` function has exponential time complexity? Let's make sure the algorithm is good before adding the cache.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: Ya. It's possible. But, if we create a Guava loading cache and pass through all the cache management logic in Guava, it means a more overhead than this PR. In this PR, `spark.sql.orc.cache.sarg.timeout=0` means not creating the loading cache at all.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala [hunk 3]
```diff
@@ -55,19 +59,52 @@ import org.apache.spark.sql.types._
...
 
     for {
       // Combines all convertible filters using `And` to produce a single conjunction
-      conjunction <- convertibleFilters.reduceOption(And)
+      conjunction <- convertibleFilters.reduceOption { (x, y) =>
+        val newFilter = org.apache.spark.sql.sources.And(x, y)
+        if (cacheExpireTimeout > 0) {
+          // Build in a bottom-up manner
+          getOrBuildSearchArgumentWithNewBuilder(dataTypeMap, newFilter)
+        }
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: Final conjunction? All sub function results will be cached in the end.

**Comment by dongjoon-hyun** [comment]
Thank you for review and advice, @cloud-fan . It turns out that my initial assessment is not enough.

First of all, from the beginning, [SPARK-2883](https://github.com/apache/spark/commit/aa31e431fc09f0477f1c2351c6275769a31aca90#diff-6cac9bc2656e3782b0312dceb8c55d47R75) is designed as a recursive function like the following. Please see `tryLeft` and `tryRight`. It's a purely computation to check if it succeeds. There is no reuse here. So, I tried to cache the first two `tryLeft` and `tryRight` operation since they can be re-used.
```scala
val tryLeft = buildSearchArgument(left, newBuilder)
val tryRight = buildSearchArgument(right, newBuilder)
val conjunction = for {
  _ <- tryLeft
  _ <- tryRight
  lhs <- buildSearchArgument(left, builder.startAnd())
  rhs <- buildSearchArgument(right, lhs)
} yield rhs.end()
```

However, before that, `createFilter` generates the target tree with [reduceOption(And)](https://github.com/apache/spark/commit/aa31e431fc09f0477f1c2351c6275769a31aca90#diff-6cac9bc2656e3782b0312dceb8c55d47R35) as a deeply skewed tree. That was the root cause. I'll update this PR.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala [hunk 4]
```diff
@@ -383,4 +386,13 @@ class OrcFilterSuite extends OrcTest with SharedSQLContext {
       )).get.toString
     }
   }
+
+  test("SPARK-25306 createFilter should not hang") {
+    import org.apache.spark.sql.sources._
+    val schema = new StructType(Array(StructField("a", IntegerType, nullable = true)))
+    val filters = (1 to 2000).map(LessThan("a", _)).toArray[Filter]
+    failAfter(2 seconds) {
+      OrcFilters.createFilter(schema, filters)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala]: This test looks tricky... It's a bad practice to assume some code will return in a certain time. Can we just add a microbenchmark for it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala [hunk 5]
```diff
@@ -71,12 +71,24 @@ private[orc] object OrcFilters {
 
     for {
       // Combines all convertible filters using `And` to produce a single conjunction
-      conjunction <- convertibleFilters.reduceOption(org.apache.spark.sql.sources.And)
+      conjunction <- buildTree(convertibleFilters)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala]: does parquet has the same problem?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala]: In parquet, this is done as
```
filters
  .flatMap(ParquetFilters.createFilter(requiredSchema, _))
  .reduceOption(FilterApi.and)
```

can we follow it?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala]: Sure. Something like the test code in the PR description? And marked as `ignore(...)` instead of `test(...)`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala (hunk 5 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala]: For the first question, I don't think Parquet has the same issue because Parquet uses `canMakeFilterOn` while ORC is trying to build a full result (with a fresh builder) to check if it's okay or not.

For the second question, in ORC, we already did the first half(`flatMap`) to compute `convertibleFilters`, but it can change it with `filters.filter`.
```scala
val convertibleFilters = for {
    filter <- filters
    _ <- buildSearchArgument(dataTypeMap, filter, SearchArgumentFactory.newBuilder())
} yield filter
```

2. And, the second half `reduceOption(FilterApi.and)` was the original ORC code which generated a skewed tree having exponential time complexity. We need to use `buildTree`.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala]: BTW, Parquet has another issue here due to `.reduceOption(FilterApi.and)`. When I make a benchmark, Parquet seems to be unable to handle 1000 filters, @cloud-fan .

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala [hunk 6]
```diff
@@ -398,6 +398,24 @@ class FilterPushdownBenchmark extends SparkFunSuite with BenchmarkBeforeAndAfter
...
   }
+
+  test(s"Pushdown benchmark with many filters") {
+    val numRows = 1
+    val width = 500
+
+    withTempPath { dir =>
+      val columns = (1 to width).map(i => s"id c$i")
+      val df = spark.range(1).selectExpr(columns: _*)
+      withTempTable("orcTable", "patquetTable") {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala]: nit: a typo, `patquetTable`.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala]: Oh, thanks!

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala [hunk 7]
```diff
@@ -54,27 +55,27 @@ import org.apache.spark.sql.types._
...
-
-    for {
-      // Combines all convertible filters using `And` to produce a single conjunction
-      conjunction <- convertibleFilters.reduceOption(org.apache.spark.sql.sources.And)
-      // Then tries to build a single ORC `SearchArgument` for the conjunction predicate
-      builder <- buildSearchArgument(dataTypeMap, conjunction, SearchArgumentFactory.newBuilder())
-    } yield builder.build()
+    buildTree(filters.filter(buildSearchArgument(dataTypeMap, _, newBuilder).isDefined))
+      .flatMap(buildSearchArgument(dataTypeMap, _, newBuilder))
+      .map(_.build)
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala]: Sure. No problem, @cloud-fan . :)

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
The previous failures are irrelevant to this PR.
- org.apache.spark.sql.execution.streaming.HDFSMetadataLogSuite.HDFSMetadataLog: metadata directory collision
- org.apache.spark.sql.hive.client.HiveClientSuites.(It is not a test it is a sbt.testing.SuiteSelector)
- org.apache.spark.sql.hive.client.HiveClientSuites.(It is not a test it is a sbt.testing.SuiteSelector)

**Comment by dongjoon-hyun** [comment]
At this time, R failure.
```
DONE ===========================================================================
Had test warnings or failures; see logs.
```

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan . Sure. I'll update them.

**Comment by dongjoon-hyun** [comment]
Also, thank you for review, @xuanyuanking, @kiszk , @viirya , @HyukjinKwon .
